volatility3/framework/plugins/linux/lsblk.py

```diff
@@ -0,0 +1,447 @@
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Linux kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+        ]
+
+    def _generator(
+        self,
+    ):
+        """Generates the list of block devices."""
+        vmlinux = self.context.modules[self.config["kernel"]]
+        try:
+            class_kset = vmlinux.object_from_symbol("class_kset")
+        except exceptions.SymbolError:
+            class_kset = None
+        if not class_kset:
+            raise TypeError(
+                "This plugin requires the class_kset structure. This structure is not present in the supplied symbol table. This means you are either analyzing an unsupported kernel version or that your symbol table is corrupt."
+            )
+
+        block_class = vmlinux.object_from_symbol("block_class")
+
+        subsys_off = vmlinux.get_type("subsys_private").relative_child_offset("subsys")
+        kobj_off = vmlinux.get_type("kset").relative_child_offset("kobj")
+
+        if not vmlinux.has_type("hd_struct"):
+            block_device_offset = vmlinux.get_type(
+                "block_device"
+            ).relative_child_offset("bd_device")
+        else:
+            hd_struct_offset = vmlinux.get_type("hd_struct").relative_child_offset(
+                "__dev"
+            )
+            gendisk_offset = vmlinux.get_type("gendisk").relative_child_offset("part0")
+
+        for kobj in class_kset.list.to_list(
+            vmlinux.symbol_table_name + constants.BANG + "kobject", "entry"
+        ):
+            subsys_private = vmlinux.object(
+                object_type="subsys_private",
+                offset=kobj.vol.offset - kobj_off - subsys_off,
+                absolute=True,
+            )
+            if subsys_private.member("class") == block_class.vol.offset:
+                break
+
+        klist_devices = subsys_private.klist_devices
+        for device in self._device_iterator(vmlinux, klist_devices):
+            # Before v5.11, partitions are represented by hd_struct instead of block_device
+            if not vmlinux.has_type("hd_struct"):
+                block_device = vmlinux.object(
+                    object_type="block_device",
+                    offset=device.vol.offset - block_device_offset,
+                    absolute=True,
+                )
+                gendisk = block_device.bd_disk
+
+                # lsblk by default skips over ram devices
+                try:
+                    if utility.array_to_string(gendisk.disk_name).startswith("ram"):
+                        continue
+                except exceptions.InvalidAddressException:
+                    continue
+            else:
+                hd_struct = vmlinux.object(
+                    object_type="hd_struct",
+                    offset=device.vol.offset - hd_struct_offset,
```

Comment:
Again, this seems like we're working backwards from a `device` to an `hd_struct`?  A quick comment to explain that's what we're doing would be handy (it'll also help everyone make sure that what we're trying to do is what we're actually doing).